### PR TITLE
Code cleanup; debug logging.

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
Documents in code reason for #28; we can close that probably... I don't think there's really an issue with how the code works.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
With my copy of the data directory, this is the new output from a run. This is a bit long (ideally we'd not print some/all of these when running normally with improved logging).
```
Skipping empty file: regex.0.1.30--2016-06-21-04-21-09.json
Skipping empty file: issue-32062-equality-relations-complexity--2016-06-21-04-21-09.json
Overwriting hyper.0.5.0 from hyper.0.5.0--2015-05-25-14-31-35.json, dated 2015-05-25 19:52:45
Skipping empty file: helloworld--2016-06-21-04-21-09.json
Overwriting helloworld from helloworld--2015-05-25-14-31-35.json, dated 2015-05-25 19:52:45
Skipping empty file: piston-image-0.3.11--2016-06-21-04-21-09.json
Skipping empty file: regex-macros.0.1.30--2016-06-21-04-21-09.json
Skipping empty file: inflate-0.1.0--2016-06-21-04-21-09.json
Skipping empty file: rust-encoding.0.2.32--2016-06-21-04-21-09.json
Overwriting html5ever-2015-05-15 from html5ever-2015-05-15--2015-05-25-14-31-35.json, dated 2015-05-25 19:52:45
Skipping empty file: hyper.0.5.0--2016-06-21-04-21-09.json
Skipping empty file: jld-day15-parser--2016-06-21-04-21-09.json
Skipping empty file: issue-32278-big-array-of-strings--2016-06-21-04-21-09.json
Overwriting regex.0.1.30 from regex.0.1.30--2015-05-25-14-31-35.json, dated 2015-05-25 19:52:45
Skipping empty file: rustc--2016-06-21-04-21-09.json
Skipping empty file: html5ever-2015-05-15--2016-06-21-04-21-09.json
Overwriting rust-encoding.0.2.32 from rust-encoding.0.2.32--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting regex-macros.0.1.30 from regex-macros.0.1.30--2015-05-25-14-31-35.json, dated 2015-05-25 19:52:45
Overwriting libcore from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_bitflags from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_typeck from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting libflate from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_privacy from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting libsyntax from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_unicode from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting libserialize from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting libstd from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_trans from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting libtest from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_data_structures from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_lint from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting libgraphviz from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting liblog from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting libarena from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting liblibc from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_borrowck from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_driver from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting libfmt_macros from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting total from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting libterm from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librbml from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting liballoc from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_resolve from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting libgetopts from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_back from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustdoc from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librand from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting libcollections from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
Overwriting librustc_llvm from rustc--2015-05-25-15-45-57.json, dated 2015-05-25 19:52:45
12626 total files
3517 skipped files
6576 merged times
1111 bootstrap times
1422 benchmarks times
```

File: rs-backend/src/load.rs [hunk 1]
```diff
@@ -103,46 +99,54 @@ impl InputData {
...
-                                                               .unwrap());
-                } else {
-                    data_benchmarks.push(TestRun::new(date, header, times, test_name));
+                &mut data_benchmarks
+            };
+
+            // A run on the same day occurred. Merge the two data sets,
+            // overwriting previous data if the same crate was benchmarked.
+            if let Some(index) = push_to.iter().position(|run: &TestRun| run.date == date) {
+                let run = &mut push_to[index];
```
Comment (Mark-Simulacrum) [inline rs-backend/src/load.rs]: If there's a way to use `iter_mut().find(...)` here we should go for it, but I think non-lexical borrow are needed for that; I may be wrong though.

**Comment by nrc** [comment]
That output looks fine for a debug run (but not for a default run).

**Comment by Mark-Simulacrum** [comment]
I can add the logging infrastructure here (separate commit); do you want me to do that? I think it's a question of whether we consider warnings (empty files/overwritten data) worthy of showing in a default run. I think that no is the probable answer.

**Comment by nrc** [comment]
No, a separate PR is fine for logging infra.

For the last commit, I think it would be better to change the bootstrap to not merge timings, rather than changing benchmarks to merge (do I have that the right way around? Anyway, not merging is correct, I think).

**Comment by Mark-Simulacrum** [comment]
Should we then drop the non-merged data sets? I'm concerned that pushing them into the array will cause bugs in the binary search code (it'll work, but possibly incorrectly)... I can't point to a specific bug though.

**Comment by Mark-Simulacrum** [comment]
> Anyway, not merging is correct, I think

For both bootstrap and benchmarks? The previous behavior (JS and Rust) was merging for benchmarks, not merging for bootstrap.

**Comment by nrc** [comment]
I hope the binary search would still work, at the least it should return *some* data set for the correct day, which I think is good enough, at least for now. It would be good to know, I think.

> For both bootstrap and benchmarks?

Yeah, I think the merging behaviour is a bug, for sure, the two should do the same thing

**Comment by Mark-Simulacrum** [comment]
> I hope the binary search would still work, at the least it should return some data set for the correct day, which I think is good enough, at least for now. *It would be good to know, I think.*

The binary search will return some index.

What would be good to know? Note that we are currently merging ~6584 data sets with little overlap, so this will probably change the results...

**Comment by Mark-Simulacrum** [comment]
In fact, removing merging completely breaks the summary page: only bootstrap is shown.

Merging is seemingly required at least for benchmark...

**Comment by nrc** [comment]
Thank you!

**Pull request merged**